### ISSUE TITLE
Simplify ReplicatedMergeTree (RFC)

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#13978]
Goals:

Lower the number of nodes and amount of coordination in ZooKeeper. Simplify coordination as much as possible.

Allow (under specific setting) inserts to replicas even if they are disconnected from ZooKeeper.

Prefer to use internal requests between replicas for notifications and querying state instead of using ZooKeeper requests or watches.

Parallel quorum inserts.

Manual duplicate inserts to different replicas should work seamlessly - as if data is replicated.
    
    
Loosely related goals:

Atomic inserts in multiple partitions.

Atomic inserts of more than one block.

Allow implementation of point update/delete.

Parallel download of parts from multiple replicas.
Bandwidth control while downloading parts.
Stop and resume unfinished merging or fetching.
Large merges/fetches should not prevent small merges/fetches to work.

Open up the possibility to move parts between shards.
   
   
Proposal:

Get rid of separate "parts" set of different replicas. There will be only one, common set of parts and all replicas will converge to that set.

Get rid of "log" and "queue". Simple inserts, merges and drops will appear implicitly as modifications to common set of "parts". Specific merge instructions: mutations, optimize deduplicate - will be written to smaller "logs" inside parts (instructions on how the larger part should be created).
   
   
Implementation:

`table_path/parts/partition_id/part_name`
\- set of parts common to all replicas.
`table_path/parts/partition_id/deduplication/checksum`
\- checksums for deduplication of last blocks.

Partitions should be separate directories for slightly better scalability.

Part can be represented by persistent or ephemeral nodes.
Ephemeral node is an intent to create part. It used when:
\- part is being written by INSERT but not committed yet;
\- part should be written with quorum but it's not satisfied yet.
Merge intent is created as persistent node, so all replicas can start merging and the intent won't disappear if first replica become offline.

Part may contain or may not contain checksum. It does not contain checksum if it represents merge intent but no replicas has been finish this merge. When first replica will finish merging, it will update node with checksum. Other replicas will validate checksums.

If part is created with merge/mutation, it will contain: - the list of parts from which it's merged along with their checksums; - instruction of what's specific about this merge (deduplicate, update/delete). TODO (what depth of merges we should store?)

Merge or mutation may fail. When exception is thrown during merge or mutation and no replicas has finished merging (checksum is unavailable), we can delete the node and restore previous nodes in its place. If other replicas finish merging but the node is deleted, merged part is also deleted locally.

Along with checksum, part contains refcount - the number of replicas successfully written this part (or: the list of all replicas successfully written this part). It is used for quorum inserts. This refcount does not guarantee that replica has the specific part at this moment of time because replica may become unavailable or lose this part on local filesystem (it only indicates that the part has been successfully written once). When the refcount for part that is written with quorum, becomes large enough, part is converted from ephemeral to persistent.

The node for partition contains sequential block number. When we create a node for new (inserted) part, we increment block number in the same transaction. Note: block numbers can contain holes.

When ZooKeeper is unavailable, we can still do INSERTs but cannot allocate common sequential block numbers. We can allocate local block numbers, write parts to separate "uncommitted" (or: "offline") directory - these parts will be available for SELECTs and merges and can be committed as new parts later, when ZooKeeper will become available.
 
 
How to do INSERT:

\- split block by partitions;
For each partition:
\- allocate local block number;
\- write part to temporary directory in "uncommitted", obtain checksum;
\- create partition node in ZooKeeper if it looks like the first insert in that partition;
For all written parts in partitions at once:
\- commit parts in memory and rename directories in local filesystem;
\- atomically: allocate block numbers in ZooKeeper; create new nodes in `parts` - either persistent or ephemeral for quorum inserts; create deduplication nodes (if duplicate, remove some elements from transaction or find existing nodes and update refcounts for them, and repeat modified transaction);
\- if zookeeper is unavailable and offline inserts are not permitted, rollback commits on local filesystem and memory (if server will die, it will rollback on restart; parts on local filesystem should contain info if they were written with offline inserts allowed);
\- add information to notification queue for other replicas so they will be notified about insert.

Similar logic is applied for: ATTACH or REPLACE partition, committing of uncommitted parts that were written offline.
   
   
How to do merge:

Every replica can select parts to merge either from uncommitted parts or from committed parts that are available in local filesystem. Replica will obtain a set of parts in relevant partition from ZooKeeper, keep the version of partition node, and only select parts to merge that does not contain ephemeral parts or unwritten parts (parts without checksums) in the middle. Then it atomically: - checks that the version of partition node did not change; deletes source nodes and creates merged node with instructions on how to merge.

If replica has all parts to perform merge (e.g. it has initiated merge) it will perform it. Otherwise it will either download required parts or download already merged part or just wait for merge to be performed by other replica.

If merged part appeared to be empty (all rows are deleted during merge), we still keep it as empty part in ZooKeeper but should not keep empty part in local dataset (and should not report the part in system.parts).

Variants: maybe keep old parts in ZK before merge is finished by one/a few replicas?
   
   
How to fetch data from replicas:

Every replica subscribes to changes of common set of parts in ZooKeeper with watches (TODO what if there are too many partitions?) or being notified by other replicas about changes and requests current state from ZooKeeper after being notified (or simply use these changes?).

It compares common set of data parts with local set of data parts (we assume relatively small number of data parts but it's tricky because it still implies O(n^2) actions) or being notified only about changes by other replicas and figures out: what data parts should be fetched or merged and what data parts should be deleted.

Merge or download is performed, checksums are compared, refcounts updated, quorum parts changed from ephemeral to persistent, etc...

When data part should be deleted, replica shouldn't delete local data immediately but keep it as inactive parts for a while.
   
   
How to drop partition:

Implement it as a special merge operation that will delete all data instead of merge. Deleted parts will be replaced to single merged part (with instruction that it's delete). It will be processed in the same way as a merge/mutation that has deleted all the records.

We should not store empty part in local dataset but should keep it in ZooKeeper for a while. We can delete the empty part from ZooKeeper if all replicas have processed the corresponding merge operation.

Additionally, we should delete all the deduplication info that is related to the block numbers in a range of delete operation (part name is stored in deduplication info).

Delete old empty partitions from ZK in background thread. Partitions will be recreated implicitly on next INSERT if there will be any.

Bonus: now we can safely drop single parts or part range.

If there is no part in ZooKeeper but there is covered parts in local dataset - replica should move them to `unexpected` directory or throw an exception as a safety measure if amount of data is large.
   
   
How to perform mutations:

Two variants.

Old: write mutations to special path in ZooKeeper. Choose mutations to process in order and assign special merges of all parts.

New: mutation is represented as data part and mutation instruction is written to special data part. Whenever merge is assigned it should also do all mutations that are after the corresponding data parts according to their block numbers. Mutation part can participate in merge (and vanish) if all older parts are already mutated.

Special data part may contain mutation instruction as a simple expression. Or it may also contain precalculated mutation data - the data that should be deleted in a form of: row numbers in data parts (1); values of primary keys and index of row within rows with the same primary key (2); and data that should be inserted.

This mutation data may be applied on SELECTs.

**Comment by filimonov** [comment]
> "uncommitted" (or: "offline") directory

Maybe that approach can also be (re)used for https://github.com/ClickHouse/ClickHouse/issues/13574 

> a. mark moving parts on an old node as OLD, during SELECTs return to a root executor all OLD part ids that were processed during query
> b. mark moving parts on a new node as NEW and ATTACH part locally on replicas, during SELECTs return to a root executor all NEW part ids that were processed during a query

**Comment by alesapin** [comment]
> We can allocate local block numbers, write parts to separate "uncommitted" (or: "offline") directory - these parts will be available for SELECTs and merges and can be committed as new parts later when ZooKeeper will become available.

If I understand correctly, "uncommitted" parts can be merged only with other "uncommitted" parts? If so we will easily add a new big part when ZooKeeper became available. If we will merge committed and uncommitted locally than we can get different set parts.

> The node for partition contains the sequential block number
> - atomically: allocate block numbers in ZooKeeper; create new nodes in parts - either persistent or ephemeral for quorum inserts;

Using current `EphemeralLockInZooKeeper` or just incrementing number in `/partition_id` node? If second it's not clear how to make it atomic.

> - add information to notification queue for other replicas so they will be notified about insert.

Shouldn't we apply the same logic for all operations producing parts? How replica will know about new merges or mutations?

> select parts to merge that does not contain ephemeral parts

Is it possible to distinguish ephemeral nodes from persistent ones in ZooKeeper?

> Variants: maybe keep old parts in ZK before merge is finished by one/a few replicas?

I think we can connect it with parts removal from the local filesystem. When a given number of replicas will remove source parts from the local filesystem (merge succeeds and and all FS caches flushed) the last one can remove nodes from ZooKeeper as well.

> Every replica subscribes to changes of common set of parts in ZooKeeper with watches

But our goal is to get rid of specific zookeeper features like watches. Seems like the proposed solution requires some notification mechanism for replicas? For example special HTTP-handlers for each replica. But it requires all replicas to know about all other replicas. 

> New mutations algorithm: mutation is represented as data part and mutation instruction is written to special data part

Not clear how this "special part" will look like. Probably something like `partition_block_block_0`? We should mutate all parts in `partition` with lower block numbers than `block`? 
And what if we need to mutate the very old part which doesn't need to be merged? We can implement other version of `MergeSelector` which will take these "special parts" into account and assign a single merge+mutation operation.


Overall review:
Designated goals seem very useful, but I think not all of them will be achieved with the proposed solution. First of all, we still use all specific features of ZooKeeper: sequential nodes, ephemeral nodes, watches, atomicity between some operations. Secondly, It doesn't seem like coordination became much easier. We will need some mechanism for "notification queue" implementation, also we will have to perform quite a computational-heavy comparison between our parts state and parts state in ZooKeeper. Different "special parts" (mutations, drops, replacements) seems strange. Also for the current implementation, it's quite hard (but still possible) to track replicated table state because the queue is processed in parallel. In the proposed solution it seems to be almost impossible: the client cannot compare complicated parts state in ZooKeeper with local parts state.

However, the two most important goals are achieved:
1) Lower the number of nodes.
2) Allow (under specific setting) inserts to replicas even if they are disconnected from ZooKeeper.

In my dreams, I'd like to represent our RSM as some kind of graph. Data parts will be vertexes of graph and transformations (merges, drops, mutations, inserts) will be edges between them. Actually, the proposed solution is very similar to the graph, but maybe we need to store transformations explicitly? I'll try to think over the representation in ZooKeeper.

P.S. I think we can improve work with empty parts and allow offline inserts in the current implementation.

**Comment by alexey-milovidov** [comment]
> If I understand correctly, "uncommitted" parts can be merged only with other "uncommitted" parts?

Yes.

> Using current EphemeralLockInZooKeeper or just incrementing number in /partition_id node? If second it's not clear how to make it atomic.

Second. How to make it atomic: read current number, atomically commit number + 1 and check that version did not change or repeat if version changed (CAS loop).

> Shouldn't we apply the same logic for all operations producing parts? How replica will know about new merges or mutations?

Replica can watch for parts node. But it may lead to O(n^2) data exchange. Direct notifications can be used for optimization, e.g. don't use watches, read parts from ZooKeeper every second but also listen to direct notifications.

> Is it possible to distinguish ephemeral nodes from persistent ones in ZooKeeper?

Yes, `ephemeralOwner` in stat.

> I think we can connect it with parts removal from the local filesystem. When a given number of replicas will remove source parts from the local filesystem (merge succeeds and and all FS caches flushed) the last one can remove nodes from ZooKeeper as well.

Yes, it's tempting to do, but I think it's better not to keep old parts in ZooKeeper because there could be too many.
Also it's not Ok to remove parts from ZK after last replica has removed them, because replicas can be offline for significant amount of time or forever.

> For example special HTTP-handlers for each replica. But it requires all replicas to know about all other replicas.

Replicas register theirself as usual by persistent node in `replicas` and ephemeral `host` node inside.

> Not clear how this "special part" will look like. Probably something like partition_block_block_0? We should mutate all parts in partition with lower block numbers than block?

Yes + mutation number equal to its block number.

> And what if we need to mutate the very old part which doesn't need to be merged?

Mutation does not necessarily imply merge: old part can be merged "with itself" and mutated.

> all specific features of ZooKeeper: sequential nodes

No, we don't use sequential nodes.
But ephemeral nodes and multi-transactions are inevitable.

**Comment by filimonov** [comment]
> For all written parts in partitions at once:
> - commit parts in memory ....
> - if zookeeper is unavailable and offline inserts are not permitted, rollback commits

It looks like there is an 'dirty read' / transaction isolation issue here.
I.e. it sounds like part will be available for selects before will archive quorum.
It sounds we need to distinguish 'dirty' / 'local only' parts while waiting for zookeeper transaction / insert quorum.

**Comment by alexey-milovidov** [comment]
> It looks like there is an 'dirty read' / transaction isolation issue here.
I.e. part will be available for selects before quorum will be archived for that.

These parts can be filtered on select with settings similar to `select_sequential_consistency` setting.

**Comment by filimonov** [comment]
> Get rid of separate "parts" set of different replicas. 

That should open the natural way to solve: https://github.com/ClickHouse/ClickHouse/issues/10368

**Comment by filimonov** [comment]
>>   Shouldn't we apply the same logic for all operations producing parts? How replica will know about new merges or mutations?
>
> Replica can watch for parts node. But it may lead to O(n^2) data exchange. 

That part can be problematic. Users tend to misuse partitioning and to create an enormous number of parts - for example, a tiny table with a lot of replicas accepting a lot of inserts (they are small so merges are almost instant, and they don't reach too many parts) and/or mutations touching all parts in the table with thousands of parts. 

Maybe we can still keep a log of the last N transactions? This way if the replica is not far behind it can watch that log, and it if's obsolete - go into the 'slow path' and compare the state of parts.

**Comment by alexey-milovidov** [comment]
This task is postponed in favor of another task.

**Comment by azat** [comment]
@alexey-milovidov can you tell which one in particular?

**Comment by alexey-milovidov** [comment]
`SharedMergeTree` - it uses the same principle as in this task but with only shared storage support.

**Comment by azat** [comment]
But this is somehow different, it is related only for shared storage.

**Comment by alexey-milovidov** [comment]
Yes, that's why this task is postponed. However, it is still the preferred way to simplify ReplicatedMergeTree. Maybe we should return to this task later.

**Comment by azat** [comment]
Then maybe it worth to keep open?

**Comment by EmeraldShift** [comment]
Is there any plan to re-visit this RFC, now that SharedMergeTree is done and being used in production?

**Comment by alexey-milovidov** [comment]
Yes, this task still makes sense - ReplicatedMergeTree can be much simpler.
